railgun/hack/generators/manifests/main.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+// -----------------------------------------------------------------------------
+// Vars
+// -----------------------------------------------------------------------------
+
+var (
+	// directory that will be used for walking the filesystem to find CRDs.
+	directory string
+
+	// patchfile is the patch that will be used for each CRD file
+	patchfile = "config/crd/patches/upgrade_compat.yaml"
+)
+
+// -----------------------------------------------------------------------------
+// Main
+//
+// This script exists as a workaround for some problems that arouse between the
+// apiextensions.k8s.io/v1beta1 and apiextensions.k8s.io/v1 versions of the
+// CustomResourceDefinition API.
+//
+// In the v1beta1 version of the API a field called "preserveUnknownFields" was
+// used to trigger pruning for API fields that are sent in the request payload
+// but not a valid part of the API. In v1 this field is effectively deprecated
+// and MUST be set to false:
+//
+//  https://github.com/kubernetes/kubernetes/pull/93078
+//
+// Problematically: even though validation enforces the "false" setting of this
+// deprecated attribute for v1 CRDs, the previous default was "true" in v1beta1
+// and if you try to apply a v1 CRD that does NOT have this field set where a
+// v1beta1 version of the CRD already exists on the cluster this can result in
+// and error and fail to upgrade the CRD.
+//
+// This program accounts for this by applying "preserveUnknownFields: false" to
+// any v1 CustomResourceDefinition to work around the upstream backwards
+// incompatibility issue automatically.
+//
+// NOTE: When this was first written using controller-gen configuration options
+//       and kubebuilder flags on the API types was attempted but none of these
+//       things could produce the right result.
+//
+//       The controller-gen CLI can't really do it (easily) because the default
+//       behavior of the marshaller in sigs.k8s.io/yaml is to not emit optional
+//       boolean fields to YAML if they're set to false. 😑
+//
+//       It's quite possible there's a better way to do this and if you find one
+//       we encourage you to put in a PR and replace this.
+//
+// -----------------------------------------------------------------------------
+
+func main() {
+	// handle arguments
+	flag.StringVar(&directory, "directory", ".", "which directory to find CRDs in")
+	flag.Parse()
+
+	// find all the YAML CRDs in the provided directory and patch them
+	if err := filepath.Walk(directory, processFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Error processing files in %s: %s\n", directory, err)
```

Comment:
```suggestion
		fmt.Fprintf(os.Stderr, "Error processing files in %s: %w\n", directory, err)
```

Reply:
In this case the error wrapping is not necessary because this error has no caller in our codebase: it can't be captured it's being printed straight to STDERR before we exit the program.

Reply:
Well, print an object error into string would lose the log meaning in this case. maybe you want err.String ?

Reply:
With an `error` type object there's no need to explicitly call `String` when using the string formatter `%s`, this will not mangle the error contents.

Example: https://play.golang.org/p/tLUs1QoGdhA

Reply:
convert a struct to string makes the print out not like  a meaningful string. thats the awkard part. 
It is not a blocker to merge.

Reply:
`error` is not a struct, it is an interface type, and the examples above show that this will result in a meaningful string. Since you've pointed out this is a non-blocker I will resolve it, let me know if there's something I've missed though.